[ci] Ubuntu package servers down / degraded

## Description

Starting in late April 2026, LightGBM's CI was significantly impacted by unavailability and degraded performance on various services involved in operations like `apt update` and `apt install`.

This is mostly out of our control, but opening this issue to track and link things to.

## Reproducible example

Sometimes, outright failures like this:

```text
Err:7 https://ppa.launchpadcontent.net/git-core/ppa/ubuntu jammy/main amd64 git-man all 1:2.54.0-0ppa1~ubuntu22.04.1
  Could not connect to ppa.launchpadcontent.net:443 (185.125.190.80), connection timed out
Err:8 https://ppa.launchpadcontent.net/git-core/ppa/ubuntu jammy/main amd64 git amd64 1:2.54.0-0ppa1~ubuntu22.04.1
  Unable to connect to ppa.launchpadcontent.net:443:
E: Failed to fetch https://ppa.launchpadcontent.net/git-core/ppa/ubuntu/pool/main/g/git/git-man_2.54.0-0ppa1%7eubuntu22.04.1_all.deb  Could not connect to ppa.launchpadcontent.net:443 (185.125.190.80), connection timed out
Fetched 8073 kB in 37s (216 kB/s)
E: Failed to fetch https://ppa.launchpadcontent.net/git-core/ppa/ubuntu/pool/main/g/git/git_2.54.0-0ppa1%7eubuntu22.04.1_amd64.deb  Unable to connect to ppa.launchpadcontent.net:443:
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
Error: Process completed with exit code 100.
```

Then later, after https://status.canonical.com/ seemed to show things as "operational", timeouts like this:

```text
Processing triggers for dbus (1.12.20-2ubuntu4.1) ...
Traceback (most recent call last):
  File "/usr/bin/add-apt-repository", line 364, in <module>
    sys.exit(0 if addaptrepo.main() else 1)
  File "/usr/bin/add-apt-repository", line 347, in main
    shortcut = handler(source, **shortcut_params)
  File "/usr/lib/python3/dist-packages/softwareproperties/shortcuts.py", line 40, in shortcut_handler
    return handler(shortcut, **kwargs)
  File "/usr/lib/python3/dist-packages/softwareproperties/ppa.py", line 82, in __init__
    if self.lpppa.publish_debug_symbols:
  File "/usr/lib/python3/dist-packages/softwareproperties/ppa.py", line 120, in lpppa
    self._lpppa = self.lpteam.getPPAByName(name=self.ppaname)
  File "/usr/lib/python3/dist-packages/lazr/restfulclient/resource.py", line 592, in __call__
    response, content = self.root._browser._request(
  File "/usr/lib/python3/dist-packages/lazr/restfulclient/_browser.py", line 429, in _request
    raise error
lazr.restfulclient.errors.ServerError: HTTP Error 504: Gateway Time-out
```

## Environment info

N/A

## Additional Comments

Some resources:

* major incident: https://status.canonical.com/#/incident/KNms6QK9ewuzz-7xUsPsNylV20jEt5kyKsd8A-3ptQEHpOd8VQ40ZQs-KD81fboQXeGZB94okNHdHBGlCv58Sw==
* https://www.reddit.com/r/Ubuntu/comments/1t1ov6s/the_major_ubuntu_server_outage_is_over/
* https://www.reddit.com/r/Ubuntu/

--- a/.ci/conda-envs/ci-core-py39.txt
+++ b/.ci/conda-envs/ci-core-py39.txt
@@ -40,6 +40,9 @@ pytest=7.4.*
 # https://github.com/lightgbm-org/LightGBM/issues/6990
 graphite2=1.3.14=*_0
 
+# starting some time around 2026, 'pip' is no longer guaranteed to be installed
+pip=25.2.*
+
 # other recursive dependencies, just
 # pinned here to help speed up solves
 bokeh=2.4.*

--- a/.ci/conda-envs/ci-core.txt
+++ b/.ci/conda-envs/ci-core.txt
@@ -28,6 +28,9 @@ python-graphviz>=0.20.3
 scikit-learn>=1.3.2
 scipy>=1.1
 
+# starting some time around 2026, 'pip' is no longer guaranteed to be installed
+pip>=24
+
 # testing-only dependencies
 cloudpickle>=3.0.0
 psutil>=5.9.8

--- a/.github/workflows/r_package.yml
+++ b/.github/workflows/r_package.yml
@@ -178,10 +178,10 @@ jobs:
           persist-credentials: false
           submodules: true
       - name: Install pandoc
-        uses: r-lib/actions/setup-pandoc@6f6e5bc62fba3a704f74e7ad7ef7676c5c6a2590  # v2.11.4
+        uses: r-lib/actions/setup-pandoc@a51a8012b0aab7c32ef9d19bf54da93f3254335e  # v2.12.0
       - name: Install tinytex
         if: startsWith(matrix.os, 'windows')
-        uses: r-lib/actions/setup-tinytex@6f6e5bc62fba3a704f74e7ad7ef7676c5c6a2590  # v2.11.4
+        uses: r-lib/actions/setup-tinytex@a51a8012b0aab7c32ef9d19bf54da93f3254335e  # v2.12.0
         env:
           CTAN_MIRROR: https://ctan.math.illinois.edu/systems/win32/miktex
           TINYTEX_INSTALLER: TinyTeX
@@ -293,7 +293,7 @@ jobs:
           persist-credentials: false
           submodules: true
       - name: Install pandoc
-        uses: r-lib/actions/setup-pandoc@6f6e5bc62fba3a704f74e7ad7ef7676c5c6a2590  # v2.11.4
+        uses: r-lib/actions/setup-pandoc@a51a8012b0aab7c32ef9d19bf54da93f3254335e  # v2.12.0
       - name: Install system dependencies
         shell: bash
         run: |

--- a/build-python.sh
+++ b/build-python.sh
@@ -185,7 +185,7 @@ while [ $# -gt 0 ]; do
   shift
 done
 
-pip install --prefer-binary 'build>=0.10.0'
+python -m pip install --prefer-binary 'build>=0.10.0'
 
 # create a new directory that just contains the files needed
 # to build the Python-package
@@ -388,7 +388,7 @@ if test "${INSTALL}" = true; then
         PACKAGE_FILE="$(echo dist/lightgbm*.tar.gz)"
     fi
     # shellcheck disable=SC2086
-    pip install \
+    python -m pip install \
         ${PIP_INSTALL_ARGS} \
         --force-reinstall \
         --no-cache-dir \

